### PR TITLE
Standardized Check Scripts and Workers for PnPQ

### PR DESCRIPTION
**Pull request by TomShen1234**
Implementation for #31 

The check scripts pretty much works, but many parts of the Github worker and the install dependency scripts are pretty much experimental, and maybe will see some feedbacks regarding their structures. But everything should work at this point (and is tested in a test environment). 

Small known issues and remarks:

The `rye` version check has to be disabled temporarily because it fails for some reason in the Github worker, but actually running `rye sync` do work. I'll leave this to be fixed after the first round of feedback. 

I also replaced the `curl` to `wget` in the `install-rye` script because many systems I've been testing come with archaic version of `curl` which doesn't support the `--output-dir` functionality. 

Finally, instead of installing `shellcheck` and `shfmt` via `apt-get`, it's downloading the current latest version directly off Github release due to the same issues described above, where some systems ships archaic version that has missing features.

### CODE REVIEW

File: bin/install-rye.bash
Comment (auspicacious) [inline bin/install-rye.bash:84]: I guess this will work.

But you should definitely move the `source` inside the `GITHUB_ENV` check, otherwise it is just going to sow confusion. The `source` is essentially a no-op when run outside of GitHub Actions, since any changes to the environment caused by `source` will be lost as soon as this script exits.

```suggestion
# Add rye to PATH for Github Actions (persist path)
# shellcheck disable=SC2310
if is_set GITHUB_ENV; then
  source "${HOME}/.rye/env"
  # Write the new path to Github env
  # shellcheck disable=SC2154
  stdmsg "PATH=${PATH}" >>"${GITHUB_ENV}"
fi
```

File: bin/install-dependencies.bash
Comment (auspicacious) [inline bin/install-dependencies.bash:50]: I explained in another comment why writing files directly into the global `/tmp` is a bad idea.

This is a good place to use the exit trap. Essentially, make a temporary directory when the script starts running and then `rm -rf` the directory in the trap when execution is finished.

File: bin/install-dependencies.bash
Comment (auspicacious) [inline bin/install-dependencies.bash:48]: Good use of the `--directory` option! But, note well what the manpage says:

```
       -C, --directory=DIR
              Change to DIR before performing any operations.  **This option is order-sensitive**, i.e. it affects all options that follow.
```

Also, [the version of `tar` shipped with MacOS](https://ss64.com/mac/tar.html) (which is based on BSD tar, not GNU tar) only knows about `-C`, not `--directory`, so it's often safer to use `-C` here.

So, in order to make sure that all tar commands use the right working directory, and the script has a fighting chance of working cross-platform, write it like this:

```
tar -C "${temp_download_dir}" -xJf shellcheck.tar.xz
```

File: README.md
Comment (auspicacious) [inline README.md:7]: Missing a couple of prerequisites here.

File: .github/workflows/check.yml
Comment (auspicacious) [inline .github/workflows/check.yml:30]: Until the code actually passes all of the lint checks, we can put this as 
```suggestion
        ./bin/check.bash || true
```

so that it always passes.

**Comment by TomShen1234** [comment]
I knew many parts were somewhat hacky, so these feedbacks are very appreciated. I'll implement the changes

File: bin/check.bash
Comment (TomShen1234) [reply bin/check.bash:48]: I guess we could deactivate it at the end to restore the `$PATH`, but it's really not necessary here as everything afterward depends on the virtual environment.

File: bin/check.bash
Comment (TomShen1234) [reply bin/check.bash:57]: It's for the `source '${HOME}/.rye/env'` line. shellcheck cannot figure out what `$HOME` is for that line, therefore the `source-path` option is used to tell it to look in the home folder for that script. But I guess I could also just deactivate that particular check for the source line, since there's only 1 of them.

File: bin/install-dependencies.bash
Comment (TomShen1234) [reply bin/install-dependencies.bash:44]: Let me try updating my test environment to a newer version (or should I say, 'less archaic' version 😅), I think it may be due to using an outdated version of Ubuntu for testing resulted in an old version of curl being installed.

File: bin/install-dependencies.bash
Comment (TomShen1234) [reply bin/install-dependencies.bash:50]: I'll try to install these via `apt-get` again.

File: bin/install-dependencies.bash
Comment (TomShen1234) [reply bin/install-dependencies.bash:54]: I'll try that again with shellcheck, but for shfmt, `apt-get` cannot find the package `golang-mvdan-sh`, unless I'm missing something. I checked

File: bin/install-dependencies.bash
Comment (auspicacious) [inline bin/install-dependencies.bash:35]: By the way, I noticed that [the 24.04 runner image comes with curl and shellcheck preinstalled](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md).

It won't hurt performance to leave them in this `apt-get` command. `apt-get` doesn't reinstall already-installed packages unless you pass the `--reinstall` flag.

**Review by auspicacious — COMMENTED** [review]
The only issue that needs resolving before approval is to move to the `24.04` image and use `apt-get` to install `shfmt`.

**Comment by TomShen1234** [comment]
Hmmm, just did that, didn't behave as expected. Still cannot find the package.

**Comment by TomShen1234** [comment]
Looks like `sudo apt-get install shfmt -y` is the way to go on Ubuntu 24.04.

**Review by auspicacious — APPROVED** [review]
I missed the fact that the package is `golang-mvdan-sh-dev` on Ubuntu while it's `golang-mvdan-sh` on Debian. That should have been a sign it wouldn't work properly.

`-dev` packages provide whatever a developer would need to build code that requires a library. For C projects, this is mostly header files, and then the code you build would link against the prebuilt shared library provided by the corresponding non- `-dev` package. I'm not sure how this works for Go, where shared libraries are not the norm.

Anyway, thank you, I will merge this and look at fixing the reported errors soon. There's some code comments I think could be improved on this PR, but we can discuss them some other time.

**Pull request merged**
